src/commands/externalDiff.ts

```diff
@@ -0,0 +1,87 @@
+'use strict';
+import { SourceControlResourceState, Uri, window } from 'vscode';
+import { Command, Commands } from './common';
+import { CommandContext } from '../commands';
+import { GitService } from '../gitService';
+import { Logger } from '../logger';
+import { Messages } from '../messages';
+
+enum ResourceGroupType {
+    Merge,
+    Index,
+    WorkingTree
+}
+
+interface Resource extends SourceControlResourceState {
+    readonly resourceGroupType: ResourceGroupType;
+}
+
+class ExternalDiffFile {
+    constructor(public uri: Uri, public staged: boolean) {
+    }
+}
+
+export class ExternalDiffCommand extends Command {
+
+    constructor(private git: GitService) {
+        super(Commands.ExternalDiff);
+    }
+
+    protected async preExecute(context: CommandContext): Promise<any> {
+        const repoPath = await this.git.getRepoPathFromUri(undefined);
+        if (!repoPath) return Messages.showNoRepositoryWarningMessage(`Unable to open changed files`);
+
+        let files: ExternalDiffFile[];
+
+        if (context.type === 'scm-states') {
+            files = [];
+
+            context.scmResourceStates.forEach((_: Resource) => {
+                files.push(new ExternalDiffFile(_.resourceUri, _.resourceGroupType === ResourceGroupType.Index));
+            });
+
+            return this.execute(repoPath, files);
+        }
+
+        const status = await this.git.getStatusForRepo(repoPath);
+        if (status === undefined) return window.showWarningMessage(`Unable to open changed files`);
+
+        if (context.type === 'scm-groups') {
+            const staged = context.scmResourceGroups[0].id === 'index';
+
+            if (staged) {
+                files = status.files.filter(_ => _.indexStatus === 'M').map(_ => new ExternalDiffFile(_.Uri, true));
+            } else {
+                files = status.files.filter(_ => _.workTreeStatus === 'M').map(_ => new ExternalDiffFile(_.Uri, false));
+            }
+        } else {
+            files = [];
+
+            status.files.forEach(_ => {
```

Comment:
Nitpick , but unless it is a simple single line `forEach`, I favor `for (const foo of foos)` 😃